feat: add `stats/incr/nanmme`

kanikasharma-18 opened the pull request:
Resolves #5587

## Description

> What is the purpose of this pull request?

This pull request:

- Implement NaN handling in the nanmme accumulator function in stats/incr/nanmme.

- Ensures that NaN inputs are ignored without breaking the incremental moving mean error computation.

- Passes all existing test cases and maintains backward compatibility.


## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.
results of tests -

<img width="500" height="500" alt="Screenshot 2025-12-31 at 7 33 24 PM" src="https://github.com/user-attachments/assets/73461946-d05b-422b-b402-5b4dc4c280e4" />

results of benchmark -

<img width="500" height="500" alt="Screenshot 2025-12-31 at 7 34 15 PM" src="https://github.com/user-attachments/assets/4a1fb294-35b3-40b6-975a-a454a071eba6" />

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

### AI Assistance

> When authoring the changes proposed in this PR, did you use any kind of AI assistance?

-   [ ] Yes
-   [x] No

If you answered "yes" above, how did you use AI assistance?

-   [ ] Code generation (e.g., when writing an implementation or fixing a bug)
-   [ ] Test/benchmark generation
-   [ ] Documentation (including examples)
-   [ ] Research and understanding


@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

kanikasharma-18 committed:
fix(stats/incr/nanmme): handle NaN in nanmme accumulator

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/benchmark/benchmark.js:
```suggestion
bench( format( '%s::accumulator-nan', pkg ), function benchmark( b ) {
```
According to current project conventions, we are moving away from string concatenation while specifying javascript benchmark names. Instead it is recommended to use `@stdlib/string/format`. 
For more information, you can check out this reference: https://github.com/stdlib-js/stdlib/issues/8647

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/docs/repl.txt:
```suggestion
    1.0
```

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/docs/repl.txt:
```suggestion
    1.0
```

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/docs/repl.txt:
```suggestion
    0.0
```

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/package.json:
```suggestion
  "name": "@stdlib/stats/incr/nanmme",
```

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/test/test.js:
```suggestion
tape( 'accumulator computes moving mean error incrementally', function testIncrementalBehavior( t ) {
	var expected = [ 1, -1.5, 0, -3, -8/3, -2/3 ];
	var actual;
	var data = [
		[ 2, 3 ],
		[ 3, -1 ],
		[ 2, 5 ],
		[ 4, -4 ],
		[ 3, 0 ],
		[ -4, 5 ]
	];
	var acc = incrNanMme( 3 );

	for ( i = 0; i < data.length; i++ ) {
		actual = acc( data[ i ][ 0 ], data[ i ][ 1 ] );
		t.strictEqual( ulpdiff( actual, expected[i] ) <= 1, true, 'returns expected value' );
	}
	t.end();
});
```

According to latest project conventions, it is preferred to use `ulpdiff` instead of calculating tolerance values.
```javascript
var ulpdiff = require( '@stdlib/number/float64/base/ulp-difference' );
```
This applies to all the tests below too.

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/README.md:
```suggestion
> Compute a moving [mean error][mean-absolute-error] (ME) incrementally, ignoring `NaN` values.
```

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/README.md:
```suggestion
// returns ~5.33
```

nakul-krishnakumar replied:
This applies to all decimal return values across the package.

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/README.md:
```suggestion
// returns ~5.67
```

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/README.md:
```suggestion
// returns ~5.67
```

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/README.md:
This section is auto-generated using tools, it is expected to leave this section blank.

kanikasharma-18 committed:
fix(stats/incr/nanmme): fix review changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kanikasharma-18 commented:
@nakul-krishnakumar This PR is ready for review.

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/benchmark/benchmark.js:
```suggestion
* Copyright (c) 2026 The Stdlib Authors.
```
These have to be changed on all the files throughout the package.

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/docs/repl.txt:
```suggestion
    Returns an accumulator function which incrementally computes a moving
    mean error (ME), ignoring `NaN` value.
```

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/docs/repl.txt:
```suggestion
    ~-2.33
```

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/docs/repl.txt:
```suggestion
    ~-2.33
```

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/docs/repl.txt:
```suggestion
    moving mean error.
```

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/docs/repl.txt:
```suggestion
    each returned value is calculated from all provided values which are not `NaN`.
```

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/docs/repl.txt:
```suggestion
    ~-2.33
```
This applies to all decimal return values across the package.

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/docs/repl.txt:
```suggestion
    -2.33
```

nakul-krishnakumar commented on lib/node_modules/@stdlib/stats/incr/nanmme/lib/main.js:
@kgryte Could you please let me know if this error handling required?

kanikasharma-18 committed:
fix: review changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kanikasharma-18 commented:
@nakul-krishnakumar Please review the PR.

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmme/lib/main.js:
```diff
@@ -0,0 +1,91 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2018 The Stdlib Authors.
```
This is a new file but the copyright year says 2018. Since this package is being added in 2026, the copyright year should be 2026. This also applies to other new files in this PR (README.md, lib/index.js, test/test.js, examples/index.js).

```suggestion
* Copyright (c) 2026 The Stdlib Authors.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmme/docs/repl.txt:
```diff
@@ -0,0 +1,45 @@
+{{alias}}( W )
+    Returns an accumulator function which incrementally computes a moving
+    mean error (ME), ignoring `NaN` value.
```
Small typo - "value" should be plural "values" to match the pattern in other NaN-handling accumulators like `incrnanmean`.

```suggestion
    mean error (ME), ignoring `NaN` values.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmme/test/test.js:
```diff
@@ -0,0 +1,114 @@
+
+var tape = require( 'tape' );
+var ulpdiff = require( '@stdlib/number/float64/base/ulp-difference' );
+var incrNanMme = require( './../lib' );
```
The variable naming should be all lowercase to match the pattern in similar packages like `@stdlib/stats/incr/nanmean` (which uses `incrnanmean`) and `@stdlib/stats/incr/mme` (which uses `incrmme`).

```suggestion
var incrnanmme = require( './../lib' );
```

Note: You'll also need to update all usages of `incrNanMme` to `incrnanmme` throughout this test file.

Planeshifter reviewed (changes requested):
Thanks @kanikasharma-18 for your first PR! Left some comments.

kanikasharma-18 committed:
fix: changes required

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kanikasharma-18 committed:
fix: fixed changes required

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kanikasharma-18 committed:
fix: review changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kanikasharma-18 commented:
@Planeshifter I have made the required changes sir. Please review my PR.

kanikasharma-18 commented:
@Planeshifter @nakul-krishnakumar @kgryte 
Please review this PR.

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmme/lib/main.js:
```diff
@@ -0,0 +1,91 @@
+* m = accumulator();
+* // returns ~-0.33
+*/
+function incrnanmme( W ) {
+	var mean;
+	if ( !isPositiveInteger( W ) ) {
+		throw new TypeError( format( 'invalid argument. Must provide a positive integer. Value: `%s`.', W ) );
+	}
+	mean = incrmmean( W );
+	return accumulator;
+
+	/**
+	* If provided input values, the accumulator function returns an updated mean error. If not provided input values, the accumulator function returns the current mean error.
+	*
+	* @private
+	* @param {number} [x] - input value
+	* @param {number} [y] - input value
+	* @returns {(number|null)} mean error or null
+	*/
+	function accumulator( x, y ) {
+		if ( arguments.length === 0 ) {
+			return mean();
+		}
+		if ( isnan( x ) || isnan( y ) ) {
+			return mean();
+		}
+		return mean( y-x );
+	}
+}
```
The standard pattern for `nan*` packages is to wrap the base package (`incrmme`) directly rather than reimplementing the logic. This way you don't duplicate the input validation and the accumulator just adds the NaN guard. See `stats/incr/nanmstdev` or `stats/incr/nanmmaape` for the expected pattern:

```js
var isnan = require( '@stdlib/math/base/assert/is-nan' );
var incrmme = require( '@stdlib/stats/incr/mme' );

function incrnanmme( W ) {
    var mme = incrmme( W );
    return accumulator;

    function accumulator( x, y ) {
        if ( arguments.length === 0 || isnan( x ) || isnan( y ) ) {
            return mme();
        }
        return mme( x, y );
    }
}
```

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmme/test/test.js:
```diff
@@ -0,0 +1,114 @@
+
+// TESTS //
+
+tape( 'main export is a function', function testMainExport( t ) {
```
Throughout this file, test callback functions are named descriptively (e.g., `testMainExport`, `testInvalidInput`, `testReturnsFunction`, etc.). The stdlib convention is to always use `function test( t )`. Please rename all test callbacks to `test`.

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmme/test/test.js:
```diff
@@ -0,0 +1,114 @@
+	acc( 3, -5 );
+	acc( 1, 10 );
+
+	t.strictEqual( acc(), 0.5 );
```
Missing assertion message. All `t.strictEqual` calls need a message argument, typically `'returns expected value'`.

```suggestion
	t.strictEqual( acc(), 0.5, 'returns expected value' );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmme/package.json:
```diff
@@ -0,0 +1,83 @@
+{
+  "name": "@stdlib/stats/incr/nanmme",
+  "version": "0.0.0",
+  "description": "Compute a moving mean error (ME) incrementally.",
```
Description should mention NaN handling, e.g.:

```suggestion
  "description": "Compute a moving mean error (ME) incrementally, ignoring NaN values.",
```

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmme/examples/index.js:
```diff
@@ -0,0 +1,40 @@
+	v1 = ( randu()*100.0 ) - 50.0;
+	v2 = ( i%10 === 0 ) ? NaN : ( ( randu()*100.0 ) - 50.0 );
+	err = accumulator( v1, v2 );
+	console.log('%s\t%s\t%s', v1.toFixed( 3 ), ( isNaN( v2 ) ) ? 'NaN' : v2.toFixed( 3 ), ( err === null || isNaN( err ) ) ? 'NaN' : err.toFixed( 3 ));
```
Use `@stdlib/math/base/assert/is-nan` instead of the global `isNaN`.

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanmme/docs/repl.txt:
```diff
@@ -0,0 +1,45 @@
+{{alias}}( W )
```
Missing leading blank line before `{{alias}}`.

Planeshifter reviewed (changes requested):
Left another round of feedback.

Planeshifter committed:
fix: address review feedback for stats/incr/nanmme

Planeshifter commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into fix-nanmme

Planeshifter committed:
fix: correct example return values and equation tags in stats/incr/nanmme

Planeshifter committed:
fix: add parentheses around ternary condition in examples

kanikasharma-18 committed:
Merge branch 'develop' into fix-nanmme

Planeshifter reviewed (approved):
Thanks @kanikasharma-18 for opening this PR and @nakul-krishnakumar for review!
